qutip/tests/test_qip.py

```diff
@@ -0,0 +1,39 @@
+import sys
+import unittest
+
+import pytest
+
+
+class TestWithoutQip(unittest.TestCase):
+    def setUp(self):
+        self._temp_qip = None
+        if sys.modules.get("qutip_qip"):
+            self._temp_qip = sys.modules["qutip_qip"]
+        sys.modules["qutip_qip"] = None
+
+    def tearDown(self):
+        if self._temp_qip:
+            sys.modules["qutip_qip"] = self._temp_qip
+        else:
+            del sys.modules["qutip_qip"]
+
+    def test_failed_import(self):
+        # Ensure 'qutip.qip' is not imported yet
+        assert "qutip.qip" not in sys.modules
```

Comment:
Instead of this, let's use pytest properly and write two fixtures, something like:
```python
@pytest.fixture
def without_qutip_qip(monkeypatch):
     monkeypatch.delitem(sys.modules, "qutip_qip", raising=False)
     monkeypatch.delitem(sys.modules, "qutip.qip", raising=False)

@pytest.fixture
def with_qutip_qip_stub(monkeypatch):
    monkeypatch.setitem(sys.modules, "qutip_qip", QutipQipStub())
     monkeypatch.delitem(sys.modules, "qutip.qip", raising=False)
```
where `QutipQipStub` is a class we write that is a very fake `qutip_qip` module (maybe with just one or two attributes).

Reply:
I see, in this way you fake a `qutip_qip` package but don't install it. That is clever.

Reply:
Done. 
By the way, to raise the ImportError one has to set `sys.modules["qutip_qip"]=None" - deleting it causes the module to be reimported instead (which will work fine on the CI test environment because `qutip_qip` is indeed not installed, but failed locally on my environment).